sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableDestinationCoder.java

```diff
@@ -45,13 +46,20 @@ public void encode(TableDestination value, OutputStream outStream)
     }
     tableSpecCoder.encode(value.getTableSpec(), outStream);
     tableDescriptionCoder.encode(value.getTableDescription(), outStream);
+    timePartitioningCoder.encode(value.getJsonTimePartitioning(), outStream);
   }
 
   @Override
   public TableDestination decode(InputStream inStream) throws IOException {
     String tableSpec = tableSpecCoder.decode(inStream);
     String tableDescription = tableDescriptionCoder.decode(inStream);
-    return new TableDestination(tableSpec, tableDescription);
+    String jsonTimePartitioning = null;
+    try {
+      jsonTimePartitioning = timePartitioningCoder.decode(inStream);
+    } catch (IOException e) {
+      // This implies we're decoding old state that did not contain TimePartitioning. Continue.
```

Comment:
I don't think this pattern is sustainable: if the actual reason for an IOException is something else other than "this is a pipeline update", this will produce silent data loss. Decoding data with a Coder can not be best-effort - I'd rather declare that this is a known update-incompatible change, than have the chance of data loss.

I believe there should be an update-compatible way of handling this: have a parameter in the coder "does it include time partitioning" that is false by default. Note that this will also be not quite update-compatible because it will change the coder's serialVersionUID. I suppose the only way to deal with that is to figure out the current (automatically generated) serialVersionUID of this coder and declare it explicitly as having this value.

Reply:
@jkff excuse my ignorance here, I've just started using beam in my team's production pipeline. Am I correct to assume that this PR, if/when merged, would mean that I'd have to restart our pipeline as it couldn't be done via update to prevent data loss?

Reply:
Only if you're using the dynamic-destinations version. And in most cases you can use drain (rather than cancel) to prevent data loss - unless your pipeline is currently stuck in an error loop.

Reply:
@jkff sounds good. hopefully this makes it into 2.2, our team would love to have this in our production pipeline.